src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -298,43 +353,72 @@ private void OnWriteCompleted(WriteContext writeContext)
             var error = writeContext.WriteError;
 
-            if (error != null)
+            PoolWriteContext(writeContext);
+
+            if (error == null)
             {
-                _lastWriteError = new IOException(error.Message, error);
+                // _numBytesPreCompleted can temporarily go negative in the event there are
+                // completed writes that we haven't triggered callbacks for yet.
+                _numBytesPreCompleted -= bytesWritten;
 
-                // Abort the connection for any failed write.
-                _connection.Abort();
+                CompleteFinishedWrites(status);
             }
+            else if (error != null)
+            {
+                // Abort the connection for any failed write
+                // Queued on threadpool so get it in as first op.
+                _connection.Abort();
 
-            PoolWriteContext(writeContext);
-
-            // _numBytesPreCompleted can temporarily go negative in the event there are
-            // completed writes that we haven't triggered callbacks for yet.
-            _numBytesPreCompleted -= bytesWritten;
+                CancelOutstandingTasks(error);
+            }
+        }
 
+        private void CompleteFinishedWrites(int status)
+        {
             // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
             // This allows large writes to complete once they've actually finished.
             var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
             while (_tasksPending.Count > 0 &&
-                   (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                   (_tasksPending.Peek().BytesToWrite) <= bytesLeftToBuffer)
             {
-                var tcs = _tasksPending.Dequeue();
-                var bytesToWrite = (int)tcs.Task.AsyncState;
+                var waitingTask = _tasksPending.Dequeue();
+                var bytesToWrite = waitingTask.BytesToWrite;
 
                 _numBytesPreCompleted += bytesToWrite;
                 bytesLeftToBuffer -= bytesToWrite;
 
-                if (_lastWriteError == null)
+                // Dispose registration if there is one
+                waitingTask.CancellationRegistration?.Dispose();
+
+                _threadPool.Complete(waitingTask.CompletionSource);
+            }
+
+            _log.ConnectionWriteCallback(_connectionId, status);
+        }
+
+        private void CancelOutstandingTasks(Exception error)
+        {
+            _lastWriteError = error;
+
+            while (_tasksPending.Count > 0)
+            {
+                var waitingTask = _tasksPending.Dequeue();
+
+                if (waitingTask.CancellationRegistration != null)
                 {
-                    _threadPool.Complete(tcs);
+                    // Only Cancellation tokens which accept cancellation
+                    // therefore have a registration should be cancelled
```

Comment:
We should only cancel writes when we've verified the token *has* been cancelled not just that it can be cancelled.